python/tvm/relay/testing/py_converter.py

```diff
@@ -0,0 +1,616 @@
+        ref, defs = self.visit(read.ref)
+        return (ast.Attribute(ref, 'value', Load()), defs)
+
+
+    def visit_ref_write(self, write: Expr):
+        '''For writing refs, we wrap the update in a thunk
+        (returning an empty tuple to match Relay's semantics)
+        that we execute at the right time. This ensures such assignments
+        can be properly nested, since assignments are statements
+        in Python but expressions in Relay'''
+        ref, ref_defs = self.visit(write.ref)
+        val, val_defs = self.visit(write.value)
+        thunk_name = self.generate_function_name('_ref_write_thunk')
+        thunk = self.create_def(
+            thunk_name, [],
+            ref_defs + val_defs + [
+                Assign([ast.Attribute(ref, 'value', Store())], val),
+                Return(self.create_call('TupleValue', []))
+            ])
+        return (self.create_call(thunk_name, []), [thunk])
+
+
+    def visit_match(self, match: Expr):
+        '''For matches, we wrap the entire expression in a thunk
+        because it is easiest to implement them using if statements.
+        For each clause, we generate a function that checks if the
+        pattern matches. If yes, we call a function that assigns
+        the variables appropriately and invokes the clause body.'''
+        data, defs = self.visit(match.data)
+        data_var = self.generate_var_name('_match_data')
+
+        # must ensure the data clause is executed exactly once
+        thunk_body = [Assign([Name(data_var, Store())], data)]
+        for clause in match.clauses:
+            check_expr = self.create_match_check(clause.lhs, Name(data_var, Load()))
+            body_def, body_name = self.create_match_clause_body(clause.lhs, clause.rhs)
+            defs.append(body_def)
+
+            # equiv: if check(data): return body(data)
+            thunk_body.append(ast.If(
+                check_expr,
+                [Return(self.create_call(body_name, [Name(data_var, Load())]))],
+                []
+            ))
+
+        # finally if nothing matches we have a failed assert (should never happen)
+        thunk_body.append(ast.Assert(NameConstant(False), Str('Match was not exhaustive')))
+
+        thunk_name = self.generate_function_name('_match_thunk')
+        thunk_def = self.create_def(thunk_name, [], defs + thunk_body)
+        return (self.create_call(thunk_name, []), [thunk_def])
+
+
+    # these are both handled in the "call" case
+    def visit_constructor(self, _):
+        pass
+    def visit_op(self, _):
+        pass
+
+
+def to_python(expr: Expr, mod=relay.Module(), target=tvm.target.create('llvm')):
+    '''Converts the given Relay expression into a Python script (as a Python AST object).
+    For easiest debugging, import the astor package and use to_source().'''
+    converter = PythonConverter(mod, target)
+    return converter.convert(expr)
+
+
+def run_as_python(expr: Expr, mod=relay.Module(), target=tvm.target.create('llvm')):
```

Comment:
set mod to `None` and initialize it to `relay..Module()` in function body so every expression is evaluated in a fresh module. https://stackoverflow.com/a/33573262

Reply:
Ah, very good to know!